### ISSUE TITLE
App does not detect heterologous vaccination scheme (Janssen + Comirnaty) 

### ISSUE THREAD

**Issue by Sebi2020** [corona-warn-app/cwa-app-android#4334]
## Avoid duplicates
* [x] Bug is not mentioned in the [FAQ](https://www.coronawarn.app/en/faq/)
* [x] Bug is specific for Android only, for general issues / questions that apply to iOS and Android please raise them in the [documentation repository](https://github.com/corona-warn-app/cwa-documentation)
* [x] Bug is not already reported in another issue

## Technical details

- Device name: Samsung Galaxy S10
- Android version: Android 11
- App version: 2.12.2

## Describe the bug

App does not properly detect heterologous vaccination schemes. For example the STIKO recommends a second vaccination with an mRNA vaccine if someone has been vaccinated with the vaccine from Johnson & Johnson first. The information text which informs users about the STIKO recommendation did not pop up until the second vaccination certificate has been added and says that the system did not detect the second vaccination.

## Steps to reproduce the issue

1. Click on certificates (German: "Zertifikate") on the bottom side navigation panel (before & after you've added the second certificate)
![screen1](https://user-images.githubusercontent.com/1862664/140406452-1be0042d-aa9d-4e5a-804a-d1daee04d860.png)
![screen2](https://user-images.githubusercontent.com/1862664/140406450-c5bf19cd-af67-488e-a466-c283c01a6780.png)
## Expected behaviour
1. Warning should pop up before second vaccination certificate is added to the cwa
2. Warning should disappear after second vaccination certificate is added to the cwa
3. Validation check should detect, that this vaccination is the second in a heterologous vaccination scheme and therefore consider the certificate to be valid (help text implies that these vaccinations are valid immediatly)

**Comment by Ein-Tim** [comment]
@Sebi2020 

I guess the problem here is that the booster vaccination certificate is not yet valid as, due to techincal limitations (see https://github.com/Digitaler-Impfnachweis/certification-apis/discussions/198) it is not possible for the app to differentiate between a "2/2" vaccination certificate which you got for a booster vaccination and a normal "2/2" certificate which you got for your normal final vaccination dose from Biontech, AstraZeneca or Moderna.

**Comment by Sebi2020** [comment]
@Ein-Tim "and not **only** afterwards" I think it makes no sense to show the message afterwards. Regardless if it popped up before. Maybe the app is not able to figure out the validity from single certificate (like it would be the case for the CovPass Check app). But it should be able to decide from the existing certificates, if a booster vaccination has already taken place and therefore hide the message afterwards. (second expected behavior in the list). The message makes no sense, cause it indicates that there wasn't any booster vaccination afterwards.

**Comment by mlenkeit** [comment]
> I guess the problem here is that the booster vaccination certificate is not yet valid as, due to techincal limitations (see [Digitaler-Impfnachweis/certification-apis#198](https://github.com/Digitaler-Impfnachweis/certification-apis/discussions/198)) it is not possible for the app to differentiate between a "2/2" vaccination certificate which you got for a booster vaccination and a normal "2/2" certificate which you got for your normal final vaccination dose from Biontech, AstraZeneca or Moderna.

@Ein-Tim for what we call _Booster Notifications_, we always only take a look at the most recent vaccination certificate. The certificate for which the QR code is displayed or anything else is not important here.

**Comment by Ein-Tim** [comment]
@mlenkeit

Okay, thanks for your answer!

**Comment by Sebi2020** [comment]
> @Sebi2020 is the STIKO information still displayed in the app or did it disappear in the meantime? In that case, you might have been subject to a rare combination of circumstances that I can then try to explain...
It disappeared in the meantime. I would be interested in knowing what are these rare circumstances, if you would explain what these are (I have a technical background in software development).

**Comment by mlenkeit** [comment]
@Sebi2020 for technical reasons, the _Booster Notification Rules_ (how we call it) are only evaluated once a day against the most recent vaccination certificate in your Corona-Warn-App (per person, of course). So it very much looks as if in your case, these rules were evaluated before you scanned your booster certificate, but that you saw the text for the first time only after the booster certificate was registered in the Corona-Warn-App. Looking at some other reports on GitHub, I guess it's not so much of an edge case. Anyway, we are working on re-evaluating the Booster Notification Rules when a new certificate is registered, so that situations like yours don't happen again.

**Comment by dsarkar** [comment]
@Sebi2020 See also 

- **DE** https://www.coronawarn.app/de/faq/#vac_booster_jj
- **EN** https://www.coronawarn.app/en/faq/#vac_booster_jj

**Comment by Ein-Tim** [comment]
Can this issue be closed?

**Comment by Dudedex** [comment]
This issue is still a problem for a lot of people out there, because currently there is no way of verifieing the "boostered" status in germany for example

**Comment by dsarkar** [comment]
Moving documentations repo. Affects all platforms.

**Comment by Sebi2020** [comment]
Is 2/2 still considered to be a booster, if first vaccination was done with J&J? Or is 2/2 just part of the basic immunization? I'm not sure if a third vaccination as a booster is planned if the first one was J & J.

**Comment by lestofante** [comment]
Sorry if wrong issue.
I had 1 J & J, then booster with Pfizer, 89 days ago.
As far as i can understand, tomorrow my certificate will expire, as the "booster" is not valid anymore and 3 months is the limit, but i got no notification that this is going to happen.
Also the lady at the impfung center told me since it is expired, for 14 days after the new booster where i will not be able to access 2G places, so i expect that from tomorrow, and until 14 days after my next booster, i will result invalid.  
Is this gonna happen in the app?

**Comment by dsarkar** [comment]
@lestofante My understanding is that with the next release (coming soon) in your case (1/1 J&J and then a second vaccination, either 2/2 or 2/1 certificate) the app would show 2G+ status until day 90 and then 2G status. I assume that after the third vaccination 2G status is still shown in the app, however, I will cross-check that. 


Best wishes, DS

---
Corona-Warn-App Open Source Team

**Comment by Sebi2020** [comment]
> I assume that after the third vaccination 2G status is still shown in the app, however, I will cross-check that.

Shouldn't the third vaccination cause the app to show the 2G+ status?

**Comment by lestofante** [comment]
> > I assume that after the third vaccination 2G status is still shown in the app, however, I will cross-check that.
> 
> Shouldn't the third vaccination cause the app to show the 2G+ status?

yes but you can register it only after 14 days since injection, no?

**Comment by Ein-Tim** [comment]
> yes but you can register it only after 14 days since injection, no?

No

**Comment by lestofante** [comment]
> > yes but you can register it only after 14 days since injection, no?
> 
> No

i managed to register it, it appear as "Vaccination 3 of 1" and it is NOT selected as "currently used certificate", and I am not sure if that is something i should change manually

**Comment by dsarkar** [comment]
@lestofante Just to summarize: 

1. You have all three certificates in the CWA:
  - 1/1 J&J
  - 2/1 
  - 3/2

2. They have all the exact same name, i.e. you see them grouped together under one single person like this: https://www.coronawarn.app/assets/screenshots/2-18/en/android/PersonDetailsFragment_cwa_2.png 

3. What status is shown for this certificate?

**Comment by lestofante** [comment]
> You have all three certificates in the CWA?

yes, they are 1/2 (J&J), 2/2, then i have 3/1

> They have all the exact same name, i.e. you see them grouped together under one single person like this: 

yes 

> What status is shown for this certificate?

I noticed 1/2 and 2/2 have a shield with a check-mark, 3/1 has a shield half white and half dark.
If I run a Check Validity, it return valid.

**Comment by lestofante** [comment]
@mlenkeit im using latest from f-droid, it is 2.14.1.3

**Comment by Ein-Tim** [comment]
@lestofante 

This version is not yet prepared for the new booster notation.

You either have to switch to CWA or wait until there is an update for CCTG.

**Comment by lestofante** [comment]
> You either have to switch to CWA or wait until there is an update for CCTG.

whops, i though is the same codebase, my bad!

**Comment by Ein-Tim** [comment]
@lestofante 

It is, but it's not the latest version. 

cc @fynngodau

**Comment by Sebi2020** [comment]
> @lestofante Just to summarize:
> 
> 1. You have all three certificates in the CWA?
> 
> * 1/1 J&J
> * 2/1
> * 3/2

Isn't the enumeration wrong? I think it must be 1 of 1 -> 2 of 2 -> 3 of 3 or 1 of 0 (recovered without any vaccination), 2 of 1 (recovered and vaccinated afterwards) -> 3 of 2 (second vaccination after recovery)

**Comment by Ein-Tim** [comment]
@Sebi2020 No, this is the correct enumeration, but it is not 3/2 but 3/1.

See https://github.com/Digitaler-Impfnachweis/certification-apis/blob/master/Implementation.md#values-for-dn-and-sd-for-base-vaccinations-with-the-same-vaccine.

**Comment by Sebi2020** [comment]
> @Sebi2020 No, this is the correct enumeration, but it is not 3/2 but 3/1.
> 
> See https://github.com/Digitaler-Impfnachweis/certification-apis/blob/master/Implementation.md#values-for-dn-and-sd-for-base-vaccinations-with-the-same-vaccine and ff.

Interesting my certificates show 1/1, 2/2, 3/3 (J&J, Biontech, Biontech)

**Comment by Ein-Tim** [comment]
@Sebi2020 Yes the enumeration has changed on 01.02.2022 and your certificates should (according to the new enumeration) be 1/1, 2/1, 3/1.

**Comment by Sebi2020** [comment]
@Ein-Tim The last update caused the app to request a new certificate for my 2/2 vaccination certificate. Now I have the new number scheme for only this one. No offer to replace my 3/3 certificate with one, which has a 3/2 its certificate.
